Data/models/Bookmark+Sync.swift

```diff
@@ -15,14 +15,48 @@ extension Bookmark {
     /// Base order is needed to distinguish between bookmarks on different devices and platforms.
     static var baseOrder: String { return Preferences.Sync.baseSyncOrder.value }
     
-    /// Sets order for all bookmarks. Needed after user joins sync group for the first time.
+    // Errors thrown when using the javascript library for calculating sync orders are returned
+    // as 'undefined'.
+    private var isSyncOrderValid: Bool { return syncOrder != nil && syncOrder != "undefined" }
+    
+    /// syncOrder for Brave < 1.8 has to be set in order to support the new ordering mechanism.
+    /// Pre 1.8 bookmarks didn't have syncOrder set which makes it hard to calculate syncOrder for new
+    // bookmarks, especially the ones inside of folders(nested Bookmarks syncOrder should be based on
+    // its parentFolder order which we don't have in pre 1.8)
+    public class func syncOrderMigration() {
+        let context = DataController.newBackgroundContext()
+        
+        context.perform {
+            let allBookmarks = getAllBookmarks(context: context)
+            let bookmarksWithInvalidSyncOrder = allBookmarks.filter { !$0.isSyncOrderValid }
+            
+            if allBookmarks.count == bookmarksWithInvalidSyncOrder.count {
+                updateBookmarksWithNewSyncOrder(context: context)
+            }
+            
+            let allFavorites = all(where: NSPredicate(format: "isFavorite == YES"), context: context) ?? []
+            let favoritesWithInvalidSyncOrder = allFavorites.filter { !$0.isSyncOrderValid }
+            
+            if allFavorites.count == favoritesWithInvalidSyncOrder.count {
+                updateBookmarksWithNewSyncOrder(updateFavorites: true, context: context)
+            }
+            
+            DataController.save(context: context)
+        }
+    }
+    
+    /// Sets order for all bookmarks. Needed after user joins sync group for the first time,
+    // or after updating from older Brave versions(<1.8.0)
     /// Returns an array of bookmarks with updated `syncOrder`.
-    class func updateBookmarksWithNewSyncOrder(parentFolder: Bookmark? = nil,
-                                               context: NSManagedObjectContext) -> [Bookmark]? {
+    @discardableResult
+    public class func updateBookmarksWithNewSyncOrder(parentFolder: Bookmark? = nil,
```

Comment:
Does this need to be `public`? Looks like its only only change in call site is from `syncOrderMigration`